asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py

```diff
@@ -0,0 +1,758 @@
+        ]
+    ]
+
+
+def process_substation_lsoa_data(
+    substations_df: gpd.GeoDataFrame,
+    lsoa_gdf: gpd.GeoDataFrame,
+    households_df: pd.DataFrame,
+) -> pd.DataFrame:
+    """
+    Process substation and LSOA data to calculate distributed headroom per household.
+
+    Args:
+        substations_df: GeoDataFrame of substations with headroom data.
+        lsoa_gdf: GeoDataFrame of LSOA boundaries.
+        households_df: DataFrame with household counts per LSOA.
+
+    Returns:
+        DataFrame with processed data including distributed headroom per household.
+    """
+    lsoa_headroom = distribute_substation_headroom(
+        substations_df, lsoa_gdf, households_df
+    )
+    return calculate_headroom_per_household(lsoa_headroom, households_df)
+
+
+def assess_heatpump_suitability(
+    lsoa_data: pd.DataFrame, power_per_heatpump: float, voltage_factor: float = 0.95
+) -> pd.DataFrame:
+    """
+    Assess the heat pump installation capacity for LSOAs based on grid capacity.
+
+    Args:
+        lsoa_data: DataFrame with LSOA data including headroom and household counts.
+        power_per_heatpump: Power consumption per heat pump in kW.
+        voltage_factor: Factor to account for voltage drop (default 0.95).
+
+    Returns:
+        DataFrame with heat pump installation capacity assessment for each LSOA.
+    """
+    # Convert headroom from MVA to kW
+    lsoa_data["headroom_kw"] = lsoa_data["distributed_headroom"] * 1000 * voltage_factor
+
+    # Calculate maximum number of heat pumps that could be installed
+    lsoa_data["max_heatpumps"] = np.floor(lsoa_data["headroom_kw"] / power_per_heatpump)
+
+    # Calculate percentage of households that could install heat pumps
+    lsoa_data["heatpump_installation_percentage"] = (
+        lsoa_data["max_heatpumps"] / lsoa_data["household_count"] * 100
+    ).clip(upper=100, lower=0)
+
+    # Calculate excess or deficit capacity
+    lsoa_data["capacity_difference_kw"] = lsoa_data["headroom_kw"] - (
+        lsoa_data["household_count"] * power_per_heatpump
+    )
+
+    return lsoa_data
+
+
+def calculate_grid_capacity() -> pd.DataFrame:
+    """
+    Calculate the grid capacity for heat pump installations across all LSOAs.
+
+    This function performs the following steps:
+    1. Generate and process substation data from all grid operators
+    2. Load and process LSOA and household data
+    3. Distribute substation headroom to LSOAs
+    4. Assess heat pump suitability based on available capacity
+
+    Returns:
+        DataFrame with grid capacity assessment results for each LSOA
+    """
+    # Generate and process substation data
+    substations = generate_substations_df()
+    substations["firm_capacity_mva"] = substations["firm_capacity_mva"].apply(
+        _parse_capacity
+    )
+    substations["peak_demand_mva"] = substations["peak_demand_mva"].apply(
+        _parse_capacity
+    )
+
+    # Apply substation rating scaling factor
```

Comment:
What is the scaling factor for?

Reply:
This scales the substation capacities used in the calculation by a uniform factor e.g. to calculate the suitability for a 10% increase in capacity for all substations, the substation scaling factor would be 1.1.